### ISSUE TITLE
App Store Connect review failed 

### ISSUE THREAD

**Issue by DmitryM9** [LiquidPlayer/LiquidCore#108]
When checking the application in the app store, I encountered this problem.

> Non-public API usage - The app references non-public symbols in Frameworks/LiquidCore.framework/LiquidCore: _JSContextCreateBacktrace, _JSContextGroupAddHeapFinalizer, _JSContextGroupAddMarkingConstraint, _JSContextGroupClearExecutionTimeLimit, _JSContextGroupRemoveHeapFinalizer, _JSContextGroupSetExecutionTimeLimit, _JSGlobalContextSetIncludesNativeCallStackWhenReportingExceptions, _JSObjectGetGlobalContext, _JSObjectGetPrivateProperty, _JSObjectGetProxyTarget, _JSObjectSetPrivateProperty, _JSScriptCreateFromString, _JSScriptEvaluate, _JSScriptRelease, _JSStringCreateWithCharactersNoCopy, _JSSynchronousGarbageCollectForDebugging, _JSWeakCreate, _JSWeakGetObject, _JSWeakRelease. If method names in your source code match the private Apple APIs listed above, altering your method names will help prevent this app from being flagged in future submissions. In addition, note that one or more of the above APIs may be located in a static library that was included with your app. If so, they must be removed.

**Comment by ericwlange** [comment]
Oh shit.  This is going to be extremely difficult if not impossible to resolve.  Let me look into it.

**Comment by ericwlange** [comment]
I have committed a change which removes all of these private API calls.  The ramifications are as follows, but none of them should have an irreparable impact on running node.  There are still some bugs, but it is mostly working with this commit.

`_JSContextCreateBacktrace`

No discernible impact.

`_JSContextGroupAddHeapFinalizer`
`_JSContextGroupAddMarkingConstraint`
`_JSContextGroupRemoveHeapFinalizer`

These impact garbage collection and the handling of weak `Persistent` handles.  V8 allows you to capture a first-pass callback on objects that are about to be collected and offer an opportunity to save them before collection.  A second-pass callback will be called when the object is fully collected.  Without capturing incremental marking, I cannot find a way to make the first pass callback.  So there will only be a callback when the object is collected and gone.

Also, I cannot capture when garbage collection is happening in JSC, so now I call my own collector every 5 seconds as a workaround.

`_JSContextGroupClearExecutionTimeLimit`
`_JSContextGroupSetExecutionTimeLimit` 

The removal of these should have no notable impact unless there are attempts to halt execution of running javascript.  That will not be possible anymore.

`_JSGlobalContextSetIncludesNativeCallStackWhenReportingExceptions`

Removing this is a NOP.

`_JSObjectGetGlobalContext`

This was used liberally.  In cases where I could reasonably workaround it, I just replaced it with another method to get the context.  However, I was not able to find a workaround when this was used for capturing the creation context of an object.  V8 allows setting up access limitations for objects copied from one context to another.  Due to the lack of this function, I had to completely disable the security capability.  The good news is that this should not affect node.  Node uses a single context per isolate and does not share objects between contexts.

`_JSObjectGetPrivateProperty`
`_JSObjectSetPrivateProperty`

Found an effective workaround.

`_JSObjectGetProxyTarget`

Proxied `Proxy` to keep a `WeakMap` of proxy targets and this mimics the desired behavior.

`_JSScriptCreateFromString`
`_JSScriptEvaluate`
`_JSScriptRelease`

Found effective, though inefficient, workarounds.

`_JSStringCreateWithCharactersNoCopy`

Simple workaround, but less efficient.

`_JSSynchronousGarbageCollectForDebugging`

Just disabled.  Should not be in production in any case.

`_JSWeakCreate`
`_JSWeakGetObject`
`_JSWeakRelease`

V8 allows the use of externalized strings that provide a callback when the string is no longer in use.  Without these methods, it is not possible to know when a string is no longer in use.  So now all the callbacks fire at the end when the isolate is destroyed.

**Comment by ericwlange** [comment]
This should be fixed with the latest commit.  Please try:

```ruby
pod 'LiquidCore', :git => 'https://github.com/LiquidPlayer/LiquidCore.git', :commit => '2f8d903'
```

to verify that (1) it works, and (2) those symbols are no longer used.

**Comment by DmitryM9** [comment]
Sounds great. Today, I’ll check and send the applet for review.

**Comment by DmitryM9** [comment]
Preliminary reviews apple we passed. Fine!

**Comment by ericwlange** [comment]
Excellent!  There was a bug posted in #109, so it may not be quite stable yet.

**Comment by ericwlange** [comment]
This has been fixed in 0.6.2.
